[CU-fayzuc] Create database model

PrzemekRychter opened the pull request:
Initial database model

Used links:
 https://github.com/Appsilon/styleguide/wiki/mongoose-typescript-models
 https://medium.com/@agentwhs/complete-guide-for-typescript-for-mongoose-for-node-js-8cc0a7e470c1
 https://mongoosejs.com/docs/subdocs.html

PrzemekRychter committed:
initial commit

PrzemekRychter committed:
Redefiend TournamentSchema - smaller schemas were created in order to easily embedd them in our TournamentSchema. As discussed previously we will create only one model (one table, The model represents a collection of documents (table of records)) In our case this model is "Tournament" exported at the end of the file

PrzemekRychter committed:
Delete _ids where no needed

PrzemekRychter committed:
change variable name from Tournament to TournamentModel since the file is using Tournament type

eastemployeer commented on server/database/models/tournamentModel.ts:
```diff
@@ -1,2 +1,40 @@
-// import mongoose from 'mongoose';
-// import Tournament from './../../../shared/types/Tournament';
+import mongoose, { Schema, Document } from "mongoose";
+import { normalize } from "node:path";
```
Nieużywane. Linijka do usunięcia.

eastemployeer commented on server/database/models/tournamentModel.ts:
```diff
@@ -1,2 +1,40 @@
+// https://mongoosejs.com/docs/subdocs.html
+
+const UserSchema = new Schema({
+  id: String,
```
Myślałem o tym później i te konflikty o których Przemek mówiłeś wynikają z tego ze mongo tworzy zarówno _id jak i id (zauważyłem, że nie zawsze jest fizycznie w bazie ALE zawsze mozna się odwołać do _id poprzez id). Co ciekawe są one różnego typu co sprawia czasem problemy. Dlatego uważam, że to pole powinno nazywać się inaczej np. po prostu user albo user_microsoft_id

Zummek replied:
I've talked to him about this, I think `microsoftId` sound good. I suggest adding length restrictions for e.g. strings, array.
Use English in PR :D

Zummek commented:
In next task/sprint we should separate tournament model and team model. In tournament, we use few times this same team. So in future if we could change the name of the team we had to change the name in few objects.
The functionality that comes from this is to create teams not only for one tournament. We could create "permanent" team or temporary team for only one tournament.

PrzemekRychter committed:
Two models(tables) TournamentModel and TeamModel
Added alias field in User interface and UserSchema (Mateusz said it
would be nice for user to be able to set his alias )
To maintain "spojność" changed owner to String type, because if
user would like to change his alias AND he would be the owner it would
generate "niespojnosc" ;)

eastemployeer commented on server/database/models/tournamentModel.ts:
```diff
@@ -1,40 +1,38 @@
+  this.name = 'to do';
 
+  next(err);
+});
```
This middleware was just for testing. You could delete it before pushing to remote.

PrzemekRychter replied:
Got it

eastemployeer reviewed (commented):
Delete this test middleware. Everything else is fine for now. @Zummek quick review and merge.

Zummek commented on shared/types/Tournament.d.ts:
```diff
@@ -1,27 +1,28 @@
-    teamA: number | null;
-    teamB: number | null;
+    teamA: number;
+    teamB: number;
```
It is not correct. Score should be null when we didn't finish the game

PrzemekRychter replied:
Got it, but how can we know if we finished game? It will be stored as a number in database. I suggest we set  "-1" value in these fields, before playing a match. You can't score -1 points. Another solution is to change database type from Number to Mixed or String

Zummek replied:
@PrzemekRychter But we could to put here Number or Null. Is Mongodb support this?

mblaze replied:
Keep in mind that we also need to account for score reporting here. Each team must report a score and only if they match the final score should be set and the match can be concluded.

Alternatively the final score can be forcefully set by the owner of the tournament when the sides do not agree on the score or they have not provided it at all.

PrzemekRychter replied:
We can add 2 fields: `lastSetBy` which references the team and `confirmed`, e.g. member of Team A sets the score, then members of team B can confirm or reset that score. Confirmation would only be available if `lastSetBy` would referece the other team.

mblaze replied:
I thought about it for a bit and I believe we should have a `score: { a: number, b: number }` for both `sideA` and `sideB` and also one for the whole match. This would allow every team to report a score for both sides and in the case of a disagreement the tournament owner would be able to compare them and set the final one to something else, keeping the history of the reported scores which might come in handy.

When you start implementing the score reporting we should have a short call about it.

Zummek reviewed (approved):
Take note of the comments and it's ready to merge

PrzemekRychter committed:
Logic of schema is still the same
This design should ensure us convenient data manipulation.
https://medium.com/@agentwhs/complete-guide-for-typescript-for-mongoose-for-node-js-8cc0a7e470c1

mblaze commented on server/database/models/tournamentModel.ts:
```diff
@@ -1,2 +1,57 @@
+import { TeamDocument } from './teamModel';
+
+const MatchSchema = new Schema({
+  rivals: {
+    teamA: { type: Types.ObjectId, ref: 'Team' },
+    teamB: { type: Types.ObjectId, ref: 'Team' },
+  },
+  score: {
+    teamA: Number,
+    teamB: Number,
+  },
```
This model seems a bit disjointed. Is there any particular reason for such a design?

What do you think about this?
```suggestion
  sideA: {
    team: { type: Types.ObjectId, ref: 'Team' },
    score: Number,
  },
  sideB: {
    team: { type: Types.ObjectId, ref: 'Team' },
    score: Number,
  },
```

This way would also be much easier to type.

eastemployeer replied:
Admit that. + I think better way to get ObjectId type is to get it from Schema (Schema.Types.ObjectId). We discussed it earlier.

PrzemekRychter replied:
1. I agree (design) and I'm changing it to your suggestion @mblaze 
2. The difference between `Schema.Types.String and Types.String`
[https://github.com/Automattic/mongoose/issues/1671](url)
[https://stackoverflow.com/questions/65831345/what-is-the-difference-between-mongoose-objectid-and-mongoose-types-objectid](url)
[https://mongoosejs.com/docs/schematypes.html](url)

```
const schema = new Schema({
     _someId: mongoose.Schema.Types.ObjectId,
});
 
const Thing = mongoose.model('Thing', schema);

m._someId = new mongoose.Types.ObjectId;
```

mblaze commented on shared/types/Tournament.d.ts:
```diff
@@ -1,27 +1,39 @@
-import User from './User';
+import User from "./User";
+import mongoose, { Schema, Document, Types } from 'mongoose';
```
I guess that the types defined in shared are used in the "business logic". If that's correct then they should not depend on mongoose or anything related to the DB layer. A dependency from business code to infrastructural code causes many problems. One of them that comes to my mind here is "poisoning" the whole core (business) part of the application with some infrastructural quirks. IMO an implementation of a use-case should not worry about the DB stuff.

So my question would be, why is this dependency necessary? Wouldn't it work if we had just simple types here?

PrzemekRychter replied:
This file only uses `mongoose.Types`, to provide this interface flexibility with help of union type
```
teams: Array<Types.ObjectId> |  Array<Team>;
team: Team | Types.ObjectId
```
A union type describes a value that can be one of several types.
I can redefine this interface (this field to be able to store Team object or ref to Team) in TournamentModel which I will do, therefore dependecy in "SharedTypes" will no longer exist.

mblaze replied:
I know what union types are ;p I'm just wondering if we couldn't resolve references in the infrastructural (DB) layer fully and then we wouldn't have to deal with this in the business logic. Also, when it comes to `mongoose.Types.Array` and the standard `Array`, do we have to use the former for `Document` types for the whole thing to work properly? It'd be much better to not have any dependency on the DB library in these models. Even if we need to use IDs for referencing then IMO it should be some custom IDs.

mblaze replied:
Also, if you reference mongoose in the shared model it would also carry over the dependency to the frontend module which would make no sense at all...

eastemployeer replied:
The problem is that (I think) TypeScript itself, doesn't provide us with ObjectId type which we have to specify here. It has to be taken from `mongoose.Types.ObjectId`

mblaze replied:
Why can't we use a simple type in the shared model, like a `string` or a `number` or not use an ID at all in the business layer?

mblaze commented on server/database/models/teamModel.ts:
```diff
@@ -0,0 +1,20 @@
+});
+
+export interface TeamDocument extends Team, mongoose.Document {
+  members: Types.Array<User>;
```
Why are you redefining members with a different type here?

PrzemekRychter replied:
If we would use this interface we could call some mongoose specific array methods on team.members  which  are not available in normal ES Array type.
[https://docs.mongodb.com/manual/reference/operator/update-array/](url)

mblaze replied:
That's exactly the thing that I think you should not do in business code. That's what I meant by "poisoning" the whole codebase with infrastructural code.

IMO the business logic should not depend on any library that is related to the infrastructure. Let's say in a few months we decide that we want to switch to a different DB engine. With such a dependency you'd have to rewrite the whole application instead of rewriting just a small DB compatibility layer.

If you feel like you're not quite getting what I mean then we can organize a quick call about this.

PrzemekRychter replied:
Yes, we can talk about it, just tell us/me what time are you available soon.

PrzemekRychter replied:
Yes, we can talk about it, just tell us/me what time are you available soon.

mblaze replied:
Lets make it a part of the monday meeting

PrzemekRychter committed:
Added interfaces
On which we can execute mongoose/mongoDb functions.
But also functions definied by us

PrzemekRychter committed:
Simplified design
Decoupled typescript models form db models
Resolved references in the infrastructural (DB) layer

Zummek commented:
Task linked: [CU-fayzuc Create database model](https://app.clickup.com/t/fayzuc)

PrzemekRychter committed:
added repositories

PrzemekRychter committed:
Updated repositories, created dbTest function.

Zummek commented on server/database/models/TournamentModel.ts:
```diff
@@ -0,0 +1,30 @@
+    },
+  },
+  isFinished: Schema.Types.Boolean,
+  date: Schema.Types.Date,
```
isFinished looks like unnecessary redundancy. We are able to read this information from `sideA.score.a !== -1 || sideB.score.a !== -1`

From FE side we distinguish between more statuses, not only finished or not

PrzemekRychter replied:
match is finished when score is correct, scores may be filled but they can be different so owner has to override scores, after that it is finished

Zummek commented on server/database/repositories/TeamRepository.ts:
```diff
@@ -0,0 +1,27 @@
+      savedTeam = await document.save();
+    } catch (err) {
+      throw new Error(`${err.name}\n${err.message}`);
+    }
```
`try{}catch(){}` looks useless. less information = worse. You `throw` new error with incomplete information about error(only name and message).  Do You plan to do something here?

Zummek commented on shared/types/Tournament.d.ts:
```diff
@@ -1,26 +1,35 @@
-  };
+  id?: string;
+  sideA: {
+    team: Team | string; 
```
Here we should have only Team as type. Team as string(id) should be resolved by BE(database or repository).
Am I right?

PrzemekRychter replied:
Yes, i will try to do that

Zummek commented on shared/types/Tournament.d.ts:
```diff
@@ -1,26 +1,35 @@
-  participants: Team[];
-  matches: Match[];
+  ownerMId: string;
+  teams:  Array<Team | string>; 
```
```suggestion
  teams:  Array<Team>; 
```
Same remark as above

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -0,0 +1,45 @@
+  };
+
+  public static insertTeam = async (tournament: Tournament | string, team : Team | string) => {
+    let tournamentDoc : (Document & Tournament) | null;
```
```suggestion
    let tournamentDoc: TournamentModel;
```
I'm not sure, but it could work

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -0,0 +1,45 @@
+    }
+    if (tournamentDoc === null) { throw new Error('Tournament has not been found'); }
+    // Does team exist?
+    let teamDoc: (Document & Team) | null;
```
This same as above - `TeamModel` as type

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -0,0 +1,45 @@
+    return savedTournament;
+  };
+
+  public static insertTeam = async (tournament: Tournament | string, team : Team | string) => {
```
1. Maybe change name to `includeTeam`? What do You think? 
2. We should pass `teams` as second arg. `teams: Team[]` because we will always push several teams, not one. So doing less operations on DB sounds good. We will be able to push one or more teams

Zummek replied:
3.  Is inserting team necessary? We need teams in tournament when we are creating tournament bcz we need to define matches based on teams

PrzemekRychter replied:
ok i will adjust

Zummek commented on server/database/DbTests.ts:
```diff
@@ -0,0 +1,39 @@
+      name: 'Memoriał  JP', ownerMId: 'identyfikator-wlasciciela', teams: [], matches: [],
+    })
+    .then((document) => {
+      const t: Tournament = document;
```
I think `document` should return `Tournament` type, this `t: Tournament` is useless or something is bad here

PrzemekRychter replied:
I suppose I used it to cast document which type is `Tournament & mongoose.Document` to `Tournament`. 
Now there is no need for such things, all functions from Repositories will return objects with types from sharedTypes.

Zummek reviewed (changes requested):
Remove `web-app/package-lock.json` from your changes. This is not in your assignment - redundant - may cause conflicts

Zummek commented on shared/types/User.d.ts:
```diff
@@ -1,3 +1,4 @@
 export default interface User {
-  id: string;
+  mId: string;
```
Maybe change `mId` to `MSId` its more telling

Zummek commented on shared/types/User.d.ts:
```diff
@@ -1,3 +1,4 @@
 export default interface User {
-  id: string;
+  mId: string;
+  alias: string;
```
Alias should be null or string. We - users don't have to set alias

PrzemekRychter replied:
ok

PrzemekRychter replied:
can it be `alias? : string`

PrzemekRychter committed:
added  repository methods

PrzemekRychter committed:
all functions returns populated documents

PrzemekRychter committed:
added fields in user model and schema

PrzemekRychter committed:
updated repositories

PrzemekRychter committed:
updateMatch is now working

Zummek committed:
Update codeLinter CI/CD

Zummek committed:
Back changes in codeLinter, add missing package

Zummek committed:
Fix package.json - wrong path for command -.-

Zummek committed:
Fix codeLinter - again

Zummek committed:
Try to run codeLinter

Zummek committed:
Try to run codeLinter #2

Zummek committed:
Try to run codeLineter #3

Zummek committed:
Try to run codeLineter #4

Zummek committed:
Try to run codeLineter #5
